tsl/src/continuous_aggs/materialize.c

```diff
@@ -521,32 +573,142 @@ drain_invalidation_log(int32 raw_hypertable_id, List **invalidations_out)
+		};
+		range_check(invalidation_range, bucket_width);
+		return invalidation_range;
+	}
+	range_check(invalidation_range, bucket_width);
+	return invalidation_range;
+}
+
 static void
-drain_materialization_invalidation_logs(int32 cagg_id, int64 materialization_invalidation_threshold,
-										List **invalidations_out)
+materialization_invalidation_log_delete_or_cut(int32 cagg_id, InternalTimeRange invalidation_range,
+											   int64 completed_threshold)
 {
-	InvalidationScanState scan_state = {
-		.invalidations = invalidations_out,
-		.materialization_invalidation_threshold = materialization_invalidation_threshold,
-		.mctx = CurrentMemoryContext,
-	};
-	ScanKeyData scankey[1];
+	ScanIterator iterator =
+		ts_scan_iterator_create(CONTINUOUS_AGGS_MATERIALIZATION_INVALIDATION_LOG,
+								RowExclusiveLock,
+								CurrentMemoryContext);
+	iterator.ctx.index = catalog_get_index(ts_catalog_get(),
+										   CONTINUOUS_AGGS_MATERIALIZATION_INVALIDATION_LOG,
+										   CONTINUOUS_AGGS_MATERIALIZATION_INVALIDATION_LOG_IDX);
+	ts_scan_iterator_scan_key_init(
+		&iterator,
+		Anum_continuous_aggs_materialization_invalidation_log_idx_materialization_id,
+		BTEqualStrategyNumber,
+		F_INT4EQ,
+		Int32GetDatum(cagg_id));
+
+	ts_scanner_foreach(&iterator)
+	{
+		HeapTuple tuple = ts_scan_iterator_tuple(&iterator);
+		Form_continuous_aggs_materialization_invalidation_log invalidation_form =
+			((Form_continuous_aggs_materialization_invalidation_log) GETSTRUCT(tuple));
+		bool delete_entry;
+		bool modify_entry;
 
-	ScanKeyInit(&scankey[0],
-				Anum_continuous_aggs_materialization_invalidation_log_idx_materialization_id,
-				BTEqualStrategyNumber,
-				F_INT4EQ,
-				Int32GetDatum(cagg_id));
+		if (invalidation_form->lowest_modified_value >= completed_threshold)
+		{
+			modify_entry = true;
+			delete_entry = true;
+		}
+		else
+		{
+			modify_entry =
+				range_cut_invalidation_entry_left(invalidation_range,
+												  &invalidation_form->lowest_modified_value,
+												  &invalidation_form->greatest_modified_value,
+												  &delete_entry);
+			if (!delete_entry && invalidation_form->greatest_modified_value >= completed_threshold)
+			{
+				invalidation_form->greatest_modified_value =
+					int64_saturating_sub(completed_threshold, 1);
+				modify_entry = true;
+				delete_entry = (invalidation_form->lowest_modified_value >
+								invalidation_form->greatest_modified_value);
+			}
+		}
 
-	ts_catalog_scan_all(CONTINUOUS_AGGS_MATERIALIZATION_INVALIDATION_LOG /*=table*/,
-						CONTINUOUS_AGGS_MATERIALIZATION_INVALIDATION_LOG_IDX /*=indexid*/
-						,
-						scankey /*=scankey*/,
-						1 /*=num_keys*/,
-						scan_del_materialization_invalidation_tuple /*=tuple_found*/,
-						RowExclusiveLock /*=lockmode*/,
-						&scan_state /*=data*/);
+		/* if we don't need to modify this entry we wont need to modify further entries
+		 * because the scan is ordered by lowest_modify_entry ASC*/
+		if (!modify_entry)
+			break;
```

Comment:
Is this always true? What if you have values above the `completion_threshold` later on? (doesn't affect correctness, just performance)

Reply:
you are right, changed to continue